File: 46. Permutations/Solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

 var permute = function(nums) {
  let candidate = []
  let result = []
  
  dfs(nums, candidate, result)
  return result;
};

var dfs = function(nums, candidate, result) {
  if (candidate.length == nums.length) {
      result.push([...candidate])
      return
  }
  
  for (let i = 0; i < nums.length; i++) {
      if (!candidate.includes(nums[i])) {
          candidate.push(nums[i])
          dfs(nums, candidate, result)
          candidate.pop()
      }
  }
}

File: 46. Permutations/Solution.java
class Solution {
  List<List<Integer>> result = new ArrayList<>();
  List<Integer> candidate = new ArrayList<>();
  public List<List<Integer>> permute(int[] nums) {
      
      dfs(nums);
      
      return result;
  }
  
  private void dfs(int[] nums) {
      
      if (candidate.size() == nums.length) {
          result.add(new ArrayList(candidate));
          return;
      }
      
      for (int j = 0; j < nums.length; j++) {
          if (!candidate.contains(nums[j])) {
              candidate.add(nums[j]);
              dfs(nums);
              candidate.remove(candidate.size() - 1);
          }
      }
  }
}
